kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/V3ResourcesModule.java

```diff
@@ -97,33 +95,19 @@ public void dispose(ExecutorService executorService) {
   }
 
   private static final class ProducerMetricsFactory implements Factory<ProducerMetrics> {
-
-    private final Provider<KafkaRestContext> context;
-    private volatile ProducerMetrics producerMetrics;
+    private final Provider<KafkaRestConfig> config;
 
     @Inject
-    ProducerMetricsFactory(Provider<KafkaRestContext> context) {
-      this.context = requireNonNull(context);
+    ProducerMetricsFactory(Provider<KafkaRestConfig> config) {
+      this.config = requireNonNull(config);
     }
 
     @Override
     public ProducerMetrics provide() {
-      ProducerMetrics localProducerMetrics = producerMetrics;
-      if (localProducerMetrics == null) {
-        synchronized (this) {
-          localProducerMetrics = producerMetrics;
-          if (localProducerMetrics == null) {
-            producerMetrics =
-                localProducerMetrics = new ProducerMetrics(context.get().getConfig(), Time.SYSTEM);
-          }
-        }
-      }
-      return localProducerMetrics;
+      return new ProducerMetrics(config.get(), Time.SYSTEM);
```

Comment:
IIRC in the original changes, Tom had concerns that this is not being initialized only once and in a thread-safe manner with the scope binding he was using at that point. Is the current HK2 incantation ensuring that this won't be called more than once? The non-singleton factory providing a singleton instance via an instance method is pretty quirky.

Reply:
Singleton property, including the fact this will be initialized only once, is enforced by HK2. The factory itself will be instantiated twice, once to provide the instance and another to dispose of it. Since the factory has no state, it's not a problem. Factories should not have any persistent state.